drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/base_loader.rb

```diff
@@ -0,0 +1,161 @@
+###
+# Copyright 2016 - 2023 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+module HmisExternalApis::AcHmis::Importers::Loaders
+  class BaseLoader
+    attr_reader :reader, :clobber, :tracker, :table_names
+
+    def self.perform(...)
+      new(...).perform
+    end
+
+    def initialize(reader:, tracker: nil, clobber:)
+      @reader = reader
+      @clobber = clobber
+      @tracker = tracker
+      @table_names = []
+    end
+
+    protected
+
+    def supports_upsert?
+      false
+    end
+
+    def runnable?
+      clobber ? true : supports_upsert?
+    end
+
+    # 'YYYY-MM-DD HH24:MM:SS'
+    DATE_TIME_FMT = '%Y-%m-%d %H:%M:%S'.freeze
+    def valid_date?(str)
+      str =~ /\A\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}\z/
+    end
+
+    def parse_date(str)
+      return unless str
+      raise ArgumentError, "Invalid date-time format. Expected 'YYYY-MM-DD HH24:MM:SS' but got '#{str}'" unless valid_date?(str)
+
+      DateTime.strptime(str, DATE_TIME_FMT)
+    end
+
+    def cde_definition(owner_type:, key:)
+      @cache ||= {}
+      @cache[[owner_type, key]] ||= cde_definitions.find_or_create(owner_type: owner_type, key: key)
```

Comment:
I'm getting an error here that I can't figure out. I thought it might be that the User Id was different, but its the same.
```
[/tmp/s3-2023-08-22/custom-data/FederalPovertyLevel.csv:0] ActiveRecord::RecordNotUnique PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "unique_index_ensuring_one_key_per_record_type"
DETAIL:  Key (owner_type, key)=(Hmis::Hud::IncomeBenefit, federal_poverty_level) already exists.
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/custom_data_element_definitions.rb:22:in `find_or_create'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/base_loader.rb:47:in `cde_definition'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/custom_data_element_base_loader.rb:35:in `new_cde_record'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/federal_poverty_level_loader.rb:39:in `block in build_records'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/csv_file.rb:32:in `block in each'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/csv_file.rb:31:in `with_index'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/csv_file.rb:31:in `each'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/federal_poverty_level_loader.rb:25:in `map'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/federal_poverty_level_loader.rb:25:in `build_records'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/custom_data_element_base_loader.rb:12:in `perform'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/custom_data_elements_importer.rb:69:in `run_loader'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/custom_data_elements_importer.rb:44:in `block (2 levels) in run!'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/custom_data_elements_importer.rb:38:in `each'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/custom_data_elements_importer.rb:38:in `block in run!'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/custom_data_elements_importer.rb:37:in `run!
```


and the existing record was below. I got it work by deleted the old record _and_ changing its key (since uniqueness constraint considers deleted records too I guess), so probably not a big deal, but just wanted to flag it.
```
Hmis::Hud::CustomDataElementDefinition.where(key: :federal_poverty_level)
=>
[#<Hmis::Hud::CustomDataElementDefinition:0x00007f4ba37af648
  id: 14,
  owner_type: "Hmis::Hud::IncomeBenefit",
  custom_service_type_id: nil,
  field_type: "string",
  key: "federal_poverty_level",
  label: "Federal Povery Level",
  repeats: false,
  data_source_id: 21,
  UserID: "1",
  DateCreated: Thu, 10 Aug 2023 13:56:13.604071000 EDT -04:00,
  DateUpdated: Thu, 10 Aug 2023 13:56:13.604071000 EDT -04:00,
  DateDeleted: nil,
  at_occurrence: false>]
```

Reply:
Oh yeah, that makes sense. Seems like we should adjust the unique constraint to exclude deleted records. This could come up at runtime too if someone removes and re-creates a CDE

Reply:
tracked this here:
https://www.pivotaltracker.com/story/show/185894590

Reply:
Definitely. The issue was coming up even _before_ I deleted the record, though, and I didn't understand why.

Reply:
This was because of a typo in the label on the server (Federal "Povery" Level)